File: 0101-0200/0147.Insertion_Sort_List.java
//20ms



/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
public ListNode insertionSortList(ListNode head) {
  ListNode curr = head, next = null;
  // l is a fake head
  ListNode l = new ListNode(0);
  
  while (curr != null) {
    next = curr.next;
    
    ListNode p = l;
    while (p.next != null && p.next.val < curr.val)
      p = p.next;
    
    // insert curr between p and p.next
    curr.next = p.next;
    p.next = curr;
    curr = next;
  }
  
  return l.next;
}
}




//





/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        if(head==null || head.next==null)return head;
        ListNode prv = null;
        ListNode slow = head;
        ListNode fast = head;
        while(fast!=null && fast.next!=null){
            prv = slow;
            slow=slow.next;
            fast=fast.next.next;
        }
        prv.next=null;
        ListNode l1 = insertionSortList(head);
        ListNode l2 = insertionSortList(slow);
        return megreList(l1,l2);
    }
    private ListNode megreList(ListNode head1,ListNode head2){
        ListNode dummy = new ListNode(0);
        ListNode curr = dummy;
        while(head1!=null && head2!=null){
            if(head1.val<head2.val){
                curr.next = head1;
                head1=head1.next;
            }
            else{
                curr.next = head2;
                head2=head2.next;
            }
            curr=curr.next;

        }
        if(head1!=null){
            curr.next = head1;
            
        }
        if(head2!=null){
            curr.next=head2;
        }
        return dummy.next;
    }
}

File: 0101-0200/0147.Insertion_Sort_List.rs
//17ms







// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn insertion_sort_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode::new(i32::MIN));
        let mut curr = head;
        
        while let Some(mut node) = curr {
            curr = node.next.take(); 
            let mut prev = &mut dummy;

            while prev.next.is_some() && prev.next.as_ref().unwrap().val < node.val {
                prev = prev.next.as_mut().unwrap();
            }
            
            node.next = prev.next.take();
            prev.next = Some(node);
        }
        
        return dummy.next
    }
}




//









// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

impl Solution {
    pub fn insertion_sort_list(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut ll = Vec::new();
        while let Some(mut node) = head.take() {
            head = node.next.take();
            ll.push(node.val);
        }

        ll.sort();
        head = None;
        for &val in ll.iter().rev() {
            head = Some(Box::new(ListNode { val, next: head }));
        }

        head
    }
}







//












